kratos/spatial_containers/bins_dynamic.h

```diff
@@ -36,232 +52,303 @@ std::size_t TDimension,
+        mNumCells = std::distance(mitPointsBegin, mitPointsEnd);
         CalculateBoundingBox();
         CalculateCellSize(mNumCells);
         AllocateCellsContainer();
         GenerateBins();
     }
 
-    //************************************************************************
-
-    BinsDynamic( IteratorType const& PointBegin, IteratorType const& PointEnd, PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize = 1 )
-        : mPointBegin(PointBegin), mPointEnd(PointEnd)
+    /**
+     * @brief Constructor for BinsDynamic with iterators, min point, max point, and optional bucket size.
+     * @param PointBegin   Iterator pointing to the beginning of the point data.
+     * @param PointEnd     Iterator pointing to the end of the point data.
+     * @param MinPoint     The minimum point of the bounding box.
+     * @param MaxPoint     The maximum point of the bounding box.
+     * @param BucketSize   Optional parameter to specify the bucket size (default is 1).
+     */
+    BinsDynamic(IteratorType const& PointBegin, IteratorType const& PointEnd, PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize = 1)
+        : mitPointsBegin(PointBegin), mitPointsEnd(PointEnd)
     {
-        if(mPointBegin==mPointEnd)
+        if (mitPointsBegin == mitPointsEnd)
             return;
 
-        mNumCells = std::distance(mPointBegin,mPointEnd);
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            mMinPoint[i] = MinPoint[i];
-            mMaxPoint[i] = MaxPoint[i];
+        mNumCells = std::distance(mitPointsBegin, mitPointsEnd);
+        auto& r_min_point = GetMinPoint();
+        auto& r_max_point = GetMaxPoint();
+        for (SizeType i = 0; i < Dimension; i++) {
+            r_min_point[i] = MinPoint[i];
+            r_max_point[i] = MaxPoint[i];
         }
         CalculateCellSize(mNumCells);
         AllocateCellsContainer();
         GenerateBins();
     }
 
-    //************************************************************************
-
-    BinsDynamic( PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize )
+    /**
+     * @brief Constructor for BinsDynamic with min point, max point, and optional bucket size.
+     * @param MinPoint     The minimum point of the bounding box.
+     * @param MaxPoint     The maximum point of the bounding box.
+     * @param BucketSize   Optional parameter to specify the bucket size (default is 1).
+     */
+    BinsDynamic(PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize)
         : mNumCells(0)
     {
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            mMinPoint[i] = MinPoint[i];
-            mMaxPoint[i] = MaxPoint[i];
+        auto& r_min_point = GetMinPoint();
+        auto& r_max_point = GetMaxPoint();
+        for (SizeType i = 0; i < Dimension; i++) {
+            r_min_point[i] = MinPoint[i];
+            r_max_point[i] = MaxPoint[i];
         }
```

Comment:
```c++
BinsDynamic(PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize)
        : mNumCells(0), mBoundingBox(MinPoint, MaxPoint)
    {
```

Reply:
okay

Reply:
Due to the evil nature of intrussive pointers, some node constructors are removed, and I need to do some minor modification of bb class in ordetr to make it work